### PR TITLE
Stop wasting time in maintaining the legacy build system

### PR DESCRIPTION
**Pull request by kholia**
@frank-dittrich @magnumripper is this slightly better? 

While we can definitely maintain the legacy build system, why do we actually care?

### CODE REVIEW

**Comment by magnumripper** [comment]
This is more like it. But I need to ponder this for a while more.

File: src/Makefile.stub [hunk 1]
```diff
@@ -26,6 +23,6 @@ default:
 
 # Better than nothing :-/
 clean:
-	+$(MAKE) -f Makefile.legacy $@
+	echo ";)"
```
Comment (magnumripper) [inline src/Makefile.stub]: The above is not anywhere near good enough. We need to `rm *.o` at the very least.

**Comment by kholia** [comment]
@magnumripper fixed (and also rebased to current bleeding-jumbo)

**Comment by frank-dittrich** [comment]
Actually, the cutrrent build error is easy to fix.
I'm just running tests for the successful build.

**Comment by kholia** [comment]
@frank-dittrich Everything is easy (more or less) to fix. The real question is, why do it?

**Comment by jfoug** [comment]
> everything is easy (more or less) to fix. The real question is, why do it?

:+1:

**Comment by jfoug** [comment]
The ONE thing I can see from legacy, is the generic build system. I do not know if we can do that in current autoconf or not.   There may be times where generic is 'still' required.

Other than that, I see almost no purpose in the legacy build system.

**Comment by frank-dittrich** [comment]
A legacy generic build requires some changes.

After applying the following patch
```
diff --git a/src/Makefile.legacy b/src/Makefile.legacy
index 00b348d..6e81001 100644
--- a/src/Makefile.legacy
+++ b/src/Makefile.legacy
@@ -53,6 +53,8 @@ OMPFLAGS =
 # file. If these hacks do not work, comment them out and see next sections.
 ifndef RELEASE_BLD
 HAVE_LIBGMP = `find 2>&1 /usr/include /usr/local/include -name "gmp.h" | grep -q "gmp.h" && echo -DHAVE_LIBGMP`
+# pp.c currently requires sys/mmap.h
+HAVE_MMAP = `find 2>&1 /usr/include/sys -name "mmap.h" | grep -q "mmap.h" && echo -DHAVE_MMAP`
 GMP_LDFLAGS = `find 2>&1 /usr/include /usr/local/include -name "gmp.h" | grep -q "gmp.h" && echo -lgmp`
 REXGEN_FLAGS = `find 2>&1 /usr/include/librexgen/api/c /usr/local/include/librexgen/api/c -name "librexgen.h" | grep -q "librexgen.h" && echo -DHAVE_LIBREXGEN=1`
 REXGEN_LDFLAGS = `find 2>&1 /usr/lib/ /usr/local/lib/ -name "liblibrexgen*.a" | grep -q "liblibrexgen" && echo -llibrexgen`
@@ -62,6 +64,7 @@ endif
 # Uncomment the below if you have GMP library installed (faster wowsrp format)
 # in case the auto-detection does not work.
 #HAVE_LIBGMP = -DHAVE_LIBGMP
+#HAVE_MMAP = -DHAVE_MMAP
 #GMP_LDFLAGS = -lgmp
 
 # CUDA settings. Instead of "-arch sm_10" you could use sm_20 or sm_30 or
@@ -180,10 +183,10 @@ BENCH_DES_BS_OBJS_DEPEND = \
 	DES_bs_b.o
 
 BENCH_MD5_OBJS_DEPEND = \
-	MD5_fmt.o MD5_std.o
+	MD5_fmt.o MD5_std.o cryptmd5_common_plug.o
 
 BENCH_BF_OBJS_DEPEND = \
-	BF_fmt.o BF_std.o
+	BF_fmt.o BF_std.o BF_common_plug.o
 
 BENCH_OBJS = \
     $(MPIOBJ) \
diff --git a/src/bench.c b/src/bench.c
index de40810..27cd246 100644
--- a/src/bench.c
+++ b/src/bench.c
@@ -72,8 +72,10 @@
 #include "memdbg.h"
 
 #if FMT_MAIN_VERSION > 11
+#ifndef BENCH_BUILD
 static char cost_msg[128 * FMT_TUNABLE_COSTS];
 #endif
+#endif
 
 long clk_tck = 0;
 
@@ -145,6 +147,7 @@ static void bench_set_keys(struct fmt_main *format,
 	}
 }
 
+#ifndef BENCH_BUILD
 #if FMT_MAIN_VERSION > 11
 static unsigned int get_cost(struct fmt_main *format, int index, int cost_idx)
 {
@@ -164,6 +167,7 @@ static unsigned int get_cost(struct fmt_main *format, int index, int cost_idx)
 	return value;
 }
 #endif
+#endif
 
 char *benchmark_format(struct fmt_main *format, int salts,
 	struct bench_results *results)
@@ -189,14 +193,17 @@ char *benchmark_format(struct fmt_main *format, int salts,
 	void *salt, *two_salts[2];
 	int index, max, i;
 #if FMT_MAIN_VERSION > 11
+#ifndef BENCH_BUILD
 	unsigned int t_cost[2][FMT_TUNABLE_COSTS];
 	int ntests, pruned;
 #endif
+#endif
 	clk_tck_init();
 
 	if (!(current = format->params.tests)) return "FAILED (no data)";
 
 #if FMT_MAIN_VERSION > 11
+#ifndef BENCH_BUILD
 	dyna_salt_init(format);
 
 	pruned = 0;
@@ -236,6 +243,7 @@ char *benchmark_format(struct fmt_main *format, int salts,
 		return s_error;
 	}
 #endif
+#endif
 	if (!(current = format->params.tests)) return "FAILED (no data)";
 	if ((where = fmt_self_test(format))) {
 		sprintf(s_error, "FAILED (%s)\n", where);
@@ -268,15 +276,18 @@ char *benchmark_format(struct fmt_main *format, int salts,
 
 		memcpy(two_salts[index], salt, format->params.salt_size);
 #if FMT_MAIN_VERSION > 11
+#ifndef BENCH_BUILD
 		for (i = 0; i < FMT_TUNABLE_COSTS &&
 		     format->methods.tunable_cost_value[i] != NULL; i++)
 			t_cost[index][i] =
 				format->methods.tunable_cost_value[i](salt);
 #endif
+#endif
 	}
 	format->methods.set_salt(two_salts[0]);
 
 #if FMT_MAIN_VERSION > 11
+#ifndef BENCH_BUILD
 	*cost_msg = 0;
 	for (i = 0; i < FMT_TUNABLE_COSTS &&
 		     format->methods.tunable_cost_value[i] != NULL; i++) {
@@ -298,6 +309,7 @@ char *benchmark_format(struct fmt_main *format, int salts,
 		strcat(cost_msg, msg);
 	}
 #endif
+#endif
 
 /* Smashed passwords: -1001 turns into -1 and -1000 turns into 0, and
    -999 turns into 1 for benchmark length. */
@@ -688,10 +700,12 @@ AGAIN:
 #endif
 
 #if FMT_MAIN_VERSION > 11
+#ifndef BENCH_BUILD
 		if (john_main_process && benchmark_time &&
 		    *cost_msg && options.verbosity >= 3)
 			puts(cost_msg);
 #endif
+#endif
 #ifdef HAVE_MPI
 		if (mpi_p > 1) {
 			gather_results(&results_m);
diff --git a/src/formats.c b/src/formats.c
index a89f6bd..0d213d8 100644
--- a/src/formats.c
+++ b/src/formats.c
@@ -55,11 +55,13 @@ void fmt_init(struct fmt_main *format)
 	if (!format->private.initialized) {
 		double d = 0;
 
+#ifndef BENCH_BUILD
 		if (options.flags & FLG_LOOPTEST) {
 			orig_min = format->params.min_keys_per_crypt;
 			orig_max = format->params.max_keys_per_crypt;
 			orig_len = format->params.plaintext_length;
 		}
+#endif
 
 		if (!(opt = getenv("OMP_SCALE")))
 			opt = cfg_get_param(SECTION_OPTIONS, NULL,
@@ -113,11 +115,13 @@ void fmt_done(struct fmt_main *format)
 #ifdef HAVE_OPENCL
 		opencl_done();
 #endif
+#ifndef BENCH_BUILD
 		if (options.flags & FLG_LOOPTEST) {
 			format->params.min_keys_per_crypt = orig_min;
 			format->params.max_keys_per_crypt = orig_max;
 			format->params.plaintext_length = orig_len;
 		}
+#endif
 
 	}
 }
@@ -149,10 +153,12 @@ static char *longcand(struct fmt_main *format, int index, int ml)
 	static char out[PLAINTEXT_BUFFER_SIZE];
 
 	memset(out, 'A' + (index % 23), ml);
+#ifndef BENCH_BUILD
 	if (!(format->params.flags & FMT_8_BIT) ||
 	    !(format->params.flags & FMT_CASE) || pers_opts.target_enc == UTF_8)
 		memcpy(out, MAXLABEL, strlen(MAXLABEL));
 	else
+#endif
 		memcpy(out, MAXLABEL_SIMD, strlen(MAXLABEL_SIMD));
 	out[ml] = 0;
 
diff --git a/src/sph_types.h b/src/sph_types.h
index ce9f24d..50966fd 100644
--- a/src/sph_types.h
+++ b/src/sph_types.h
@@ -47,7 +47,9 @@
 #ifndef SPH_TYPES_H__
 #define SPH_TYPES_H__
 
+#if AC_BUILT
 #include "autoconfig.h"
+#endif
 
 #if !AC_BUILT || HAVE_LIMITS_H
 #include <limits.h>
```

and (temporarily moving mssql12_fmt_plug.c out of the way,
```
make -s -f Makefile.legacy generic
```
succeeds.

```
(bleeding-jumbo)src $ make -s -f Makefile.legacy generic
Compiling: DES benchmark (code version #1)
Benchmarking: descrypt, traditional crypt(3) [DES 48/64 4K]... 392704 c/s real, 392704 c/s virtual
Compiling: DES benchmark (code version #2)
Benchmarking: descrypt, traditional crypt(3) [DES 48/64 128K]... 368128 c/s real, 368128 c/s virtual
Compiling: DES benchmark (code version #3)
Benchmarking: descrypt, traditional crypt(3) [DES 48/64 4K]... 378880 c/s real, 378880 c/s virtual
Compiling: DES benchmark (code version #4)
Benchmarking: descrypt, traditional crypt(3) [DES 48/64 4K]... 376832 c/s real, 376832 c/s virtual
Compiling: DES benchmark (code version #5)
Benchmarking: descrypt, traditional crypt(3) [DES 48/64 4K]... 355072 c/s real, 355072 c/s virtual
Compiling: DES benchmark (code version #1, no key copying)
Benchmarking: descrypt, traditional crypt(3) [DES 48/64 4K]... 400640 c/s real, 400640 c/s virtual
Compiling: DES benchmark (bitslice, code version #1)
Benchmarking: descrypt, traditional crypt(3) [DES 64/64]... 1516K c/s real, 1516K c/s virtual
Compiling: DES benchmark (bitslice, code version #2)
Benchmarking: descrypt, traditional crypt(3) [DES 64/64]... 1741K c/s real, 1741K c/s virtual
Compiling: DES benchmark (bitslice, code version #3)
Benchmarking: descrypt, traditional crypt(3) [DES 64/64]... 877760 c/s real, 877760 c/s virtual
Benchmarking: md5crypt, crypt(3) $1$ [MD5 32/64]... 7789 c/s real, 7789 c/s virtual
Compiling: MD5 benchmark (two hashes at a time)
Benchmarking: md5crypt, crypt(3) $1$ [MD5 32/64 X2]... 11666 c/s real, 11666 c/s virtual
Compiling: MD5 benchmark (immediate values)
Benchmarking: md5crypt, crypt(3) $1$ [MD5 32/64 X2]... 11601 c/s real, 11718 c/s virtual
Benchmarking: bcrypt ("$2a$05", 32 iterations) [Blowfish 32/64]... 390 c/s real, 390 c/s virtual
Compiling: Blowfish benchmark (scale)
Benchmarking: bcrypt ("$2a$05", 32 iterations) [Blowfish 32/64]... 409 c/s real, 409 c/s virtual
Compiling: Blowfish benchmark (two hashes at a time)
Benchmarking: bcrypt ("$2a$05", 32 iterations) [Blowfish 32/64 X2]... 744 c/s real, 744 c/s virtual
SybaseASE_fmt_plug.c: In function ‘init’:
SybaseASE_fmt_plug.c:103:6: warning: unused variable ‘i’ [-Wunused-variable]
  int i;
      ^
pp.c: In function ‘do_prince_crack’:
pp.c:2324:5: warning: implicit declaration of function ‘munmap’ [-Wimplicit-function-declaration]
     munmap(mem_map, file_len);
     ^
ar: creating aes.a
```

And self test passes.
```
$ ./john --test=0|LC_ALL=C grep -v PASS
All 294 formats passed self-tests!
```

mssql12_fmt_plug.c still has these issues:
```
mssql12_fmt_plug.c: In function ‘clear_keys’:
mssql12_fmt_plug.c:186:44: error: ‘max_keys’ undeclared (first use in this function)
  memset(saved_key, 0, sizeof(*saved_key) * max_keys);
                                            ^
mssql12_fmt_plug.c:186:44: note: each undeclared identifier is reported only once for each function it appears in
mssql12_fmt_plug.c: At top level:
mssql12_fmt_plug.c:184:13: warning: ‘clear_keys’ defined but not used [-Wunused-function]
 static void clear_keys()
             ^
make[1]: *** [mssql12_fmt_plug.o] Error 1
make: *** [generic] Error 2
```

**Comment by frank-dittrich** [comment]
On a 64bit Linux, non-GPU build, II tested whether the autoconf build still works with my patch applied. 

Compared to an AC build, my generic build not only lacks  `mssql12` format (expected, since I removed it), but also
```
crypt
OpenVMS
Raw-SHA1-Linkedin
```

**Comment by magnumripper** [comment]
> everything is easy (more or less) to fix. The real question is, why do it?

Here's my one and only reason: Core merges. Almost every time I'm bitching about things, it's because I have core merges in mind. You have no idea how much I struggled with the 1.8 patches.

Core now has MIC patches (I presume including build targets) still to be merged. I haven't looked yet but merging that is almost certainly non-trivial. If we keep support for using the "legacy Makefile" it will help doing such merges. Then I'd first merge the changes to the legacy file, find out how to build stuff using that, and only then add whatever is needed to the autoconf stuff.

However, I'm totally fine with dropping *end user support* for legacy Makefile. We can drop all mentions of it. And/or we can spell out that if someone uses the legacy Makefile, problems are expected and you'll be on your own.

**Comment by magnumripper** [comment]
...so I think we should drop your proposed changes to sipdump and vncpcap2john. And I'm leaning towards not even moving Makefile.legacy to unused. Just keep it around, officially not supported for use by anyone not willing to duct tape it up himself. But still there.

**Comment by magnumripper** [comment]
> The ONE thing I can see from legacy, is the generic build system. I do not know if we can do that in current autoconf or not. There may be times where generic is 'still' required.

The autoconf stuff (should) totally obsolete the generic builds. The only problem is it doesn't fully do that yet. It's a problem for us that Solar isn't interested in autconf for core.

**Comment by kholia** [comment]
@magnumripper it makes sense to me now, patching according to your needs :+1:

**Pull request merged**